### PR TITLE
QLauncher data pickled with protocol 2

### PR DESCRIPTION
**Pull request by philippjfr**
As the title says, this ensures that the QLauncher data is pickled with the highest available protocol. The default protocol (0) is slow and I have frequently encountered issues pickling certain objects, particularly involving C(++) extensions and some objects using ``__slots__``. 

Defaulting to the highest available protocol would gain size/speed benefits and address the issues pickling certain complex objects with little effect on backward compatibility. In Python 2 it would use protocol 2 (introduced in Python 2.3) and in Python 3 it would use protocol 3 or 4. If we want to keep support between Python3 and Python2, we could also consider setting it to 2.

### CODE REVIEW

**Comment by jbednar** [comment]
Can you summarize the advantages of 3 or 4 over 2?  If they are minor, it would make sense to stay with 2 for cross-Python compatibility.

**Comment by philippjfr** [comment]
After looking into it I agree that protocol 2 makes the most sense. Here's the breakdown according to the pickle docs:

- Protocol version 0 is the original “human-readable” protocol and is backwards compatible with earlier versions of Python.
- Protocol version 1 is an old binary format which is also compatible with earlier versions of Python.
- Protocol version 2 was introduced in Python 2.3. It provides much more efficient pickling of new-style classes. Refer to PEP 307 for information about improvements brought by protocol 2.
- Protocol version 3 was added in Python 3.0. It has explicit support for bytes objects and cannot be unpickled by Python 2.x. This is the default protocol, and the recommended protocol when compatibility with other Python 3 versions is required.
- Protocol version 4 was added in Python 3.4. It adds support for very large objects, pickling more kinds of objects, and some data format optimizations. Refer to PEP 3154 for information about improvements brought by protocol 4.

**Comment by jbednar** [comment]
Sounds like version 2 makes the most sense, though if you noticed a huge improvement with 3 or 4 in practice that could change things.  Yes, I agree that version 2 makes sense as a better standard for holoviews, since it is interchangeable yet higher performance.

**Comment by philippjfr** [comment]
Turns out it is what we're using in HoloViews for the reference data, no longer sure why reference data can't be shared between versions. Could be that it already works and we're only doing it because the display data differs. Anyhow, protocol 2 is definitely the right choice here since it will let you launch Python 2 jobs from Python 3 and vice versa.

**Comment by jlstevens** [comment]
I forgot to make a comment here when I briefly discussed this with Philipp.

It would be *extremely* surprising if the protocol had any noticeable effect on speed. QLauncher pickles/unpickles tiny little objects (namely itself) so speed seems like poor justification to change the protocol. Perhaps you could convince me that something surprising is going on as this step should have an absolutely negligible effect on performance.

If as you say, protocol 2 is the best option for compatibility between Python 2 and 3 then that alone would be a good reason to merge this PR.

**Comment by philippjfr** [comment]
Agreed, the speed is probably a non-issue, but note that the prime reason I gave is compatibility with special types of classes such as C extensions and some classes with ``__slots__`` defined but no ``__getstate__``, which has repeatedly bitten me. Protocol 0 and 1 are also compatible with both Python2/3 but they're both ancient and there's no good reason to use them at this point.

**Comment by jlstevens** [comment]
Agreed. I'll merge it now as we do want to maximize compatibility between pickles and Python versions.
